File: java/src/main/java/com/github/koraktor/steamcondenser/steam/community/GameItem.java
/**
 * This code is free software; you can redistribute it and/or modify it under
 * the terms of the new BSD License.
 *
 * Copyright (c) 2011, Sebastian Staudt
 */

package com.github.koraktor.steamcondenser.steam.community;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import com.github.koraktor.steamcondenser.exceptions.WebApiException;

/**
 * Provides basic functionality to represent an item in a game
 *
 * @author Sebastian Staudt
 */
public class GameItem {

    private JSONArray attributes;

    private int backpackPosition;

    private String className;

    private int count;

    private int defindex;

    private int id;

    private int level;

    private String name;

    private String quality;

    private String slot;

    private boolean tradeable;

    private String type;

    /**
     * Creates a new instance of a GameItem with the given data
     *
     * @param inventory The inventory this item is contained in
     * @param itemData The data specifying this item
     * @throws JSONException on invalid JSON data
     * @throws WebApiException on Web API errors
     */
    public GameItem(GameInventory inventory, JSONObject itemData)
            throws JSONException, WebApiException {
        this.defindex         = itemData.getInt("defindex");
        this.backpackPosition = (int) itemData.getLong("inventory") & 0xffff;
        this.className        = inventory.getItemSchema().get(this.defindex).getString("item_class");
        this.count            = itemData.getInt("quantity");
        this.id               = itemData.getInt("id");
        this.level            = itemData.getInt("level");
        this.name             = inventory.getItemSchema().get(this.defindex).getString("item_name");
        this.quality          = inventory.getQualitySchema().get(itemData.getInt("quality"));
        this.slot             = inventory.getItemSchema().get(this.defindex).getString("item_slot");
        this.tradeable        = itemData.isNull("flag_cannot_trade") || !itemData.getBoolean("flag_cannot_trade");
        this.type             = inventory.getItemSchema().get(this.defindex).getString("item_type_name");

        if(!inventory.getItemSchema().get(this.defindex).isNull("attributes")) {
          this.attributes = inventory.getItemSchema().get(this.defindex).getJSONArray("attributes");
        }
    }

    /**
     * Return the attributes of this item
     *
     * @return The attributes of this item
     */
    public JSONArray getAttributes() {
        return this.attributes;
    }

    /**
     * Returns the position of this item in the player's inventory
     *
     * @return The position of this item in the player's inventory
     */
    public int getBackpackPosition() {
        return this.backpackPosition;
    }

    /**
     * Returns the class of this item
     *
     * @return The class of this item
     */
    public String getClassName() {
        return this.className;
    }

    /**
     * Returns the number of items the player owns of this item
     *
     * @return The quanitity of this item
     */
    public int getCount() {
        return this.count;
    }

    /**
     * Returns the index where the item is defined in the schema
     *
     * @return The schema index of this item
     */
    public int getDefIndex() {
        return this.defindex;
    }

    /**
     * Returns the ID of this item
     *
     * @return The ID of this item
     */
    public int getId() {
        return this.id;
    }

    /**
     * Returns the level of this item
     *
     * @return The level of this item
     */
    public int getLevel() {
        return this.level;
    }

    /**
     * Returns the level of this item
     *
     * @return The level of this item
     */
    public String getName() {
        return this.name;
    }

    /**
     * Returns the quality of this item
     *
     * @return The quality of this item
     */
    public String getQuality() {
        return this.quality;
    }

    /**
     * Returns the slot where this item can be equipped in
     *
     * @return The slot where this item can be equipped in
     */
    public String getSlot() {
        return this.slot;
    }

    /**
     * Returns whether this item is tradeable
     *
     * @return bool Whether this item is tradeable
     */
    public boolean isTradeable() {
        return this.tradeable;
    }

    /**
     * Returns the type of this item
     *
     * @return The type of this item
     */
    public String getType() {
        return this.type;
    }

}

File: ruby/lib/steam/community/game_item.rb
# This code is free software; you can redistribute it and/or modify it under
# the terms of the new BSD License.
#
# Copyright (c) 2011, Sebastian Staudt

require 'steam/community/web_api'

# A module implementing basic functionality for classes representing an item in
# a game
#
# @author Sebastian Staudt
module GameItem

  # Return the attributes of this item
  #
  # @return [Hash<Symbol, Object>] The attributes of this item
  attr_reader :attributes

  # Returns the position of this item in the player's inventory
  #
  # @return [Fixnum] The position of this item in the player's inventory
  attr_reader :backpack_position

  # Returns the class of this item
  #
  # @return [String] The class of this item
  attr_reader :class

  # Returns the number of items the player owns of this item
  #
  # @return [Fixnum] The quanitity of this item
  attr_reader :count

  # Returns the index where the item is defined in the schema
  #
  # @return [Fixnum] The schema index of this item
  attr_reader :defindex

  # Returns the ID of this item
  #
  # @return [Fixnum] The ID of this item
  attr_reader :id

  # Returns the level of this item
  #
  # @return [Fixnum] The level of this item
  attr_reader :level

  # Returns the level of this item
  #
  # @return [String] The level of this item
  attr_reader :name

  # Returns the quality of this item
  #
  # @return [String] The quality of this item
  attr_reader :quality

  # Returns the slot where this item can be equipped in
  #
  # @return [String] The slot where this item can be equipped in
  attr_reader :slot

  # Returns the type of this item
  #
  # @return [String] The type of this item
  attr_reader :type

  # Creates a new instance of a GameItem with the given data
  #
  # @param [GameInventory] inventory The inventory this item is contained in
  # @param [Hash<Symbol, Object>] item_data The data representing this item
  def initialize(inventory, item_data)
    @defindex          = item_data[:defindex]
    @backpack_position = item_data[:inventory] & 0xffff
    @class             = inventory.item_schema[@defindex][:item_class]
    @count             = item_data[:quantity]
    @id                = item_data[:id]
    @level             = item_data[:level]
    @name              = inventory.item_schema[@defindex][:item_name]
    @quality           = inventory.qualities[item_data[:quality]]
    @slot              = inventory.item_schema[@defindex][:item_slot]
    @tradeable         = !(item_data[:flag_cannot_trade] == true)
    @type              = inventory.item_schema[@defindex][:item_type_name]

    unless inventory.item_schema[@defindex][:attributes].nil?
      @attributes = inventory.item_schema[@defindex][:attributes]
    end
  end

  # Returns whether this item is tradeable
  #
  # @return [Boolean] `true` if this item is tradeable
  def tradeable?
    @tradeable
  end

end
